keps/sig-node/20200310-ensure-secret-pulled-images.md

```diff
@@ -0,0 +1,169 @@
+    - [Examples](#examples)
+      - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
+      - [Beta -&gt; GA Graduation](#beta---ga-graduation)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+We will add support for ensuring images pulled with pod imagePullSecrets are
+always authenticated even if cached. We will add a new boolean field
+`ensureSecretPulledImages` to the pod spec. The default to false
+means that if a first pod results in an image pulled with imagePullSecrets a
+second pod would have to be using always pull to ensure rights to use the
+previously pulled image. When set to true always pull would not be required,
+instead kublet will check if the image was pulled with an image pull secret and
+if so would force a pull of the image to ensure the image pulled with the
+secret is not used by another pod unless that pod also has the proper auth.
+
+## Motivation
+
+There have been customer requests for improving upon kubernetes ability to
+secure images pulled with auth. on a node. Issue
+[#18787](https://github.com/kubernetes/kubernetes/issues/18787) has been around
+for a while.
+
+To secure images one currently needs to inject `AllwaysPullImages` into pod
+specs via an admission plugin. As @liggitt [notes](https://github.com/kubernetes/kubernetes/issues/18787#issuecomment-532280931)
+the `pull` does not re-pull already-pulled layers of the image, but simply
+resolves/verifies the image manifest has not changed in the registry (which
+incidentally requires authenticating to private registries, which enforces the
+image access). That means in the normal case (where the image has not changed
+since the last pull), the request size is O(kb). However, the `pull` does put
+the registry in the critical path of starting a container, since an unavailable
+registry will fail the pull image manifest check (with or without proper
+authentication.)
+
+
+### Goals
+
+Add a flag processed by `kubelet` for `ensureSecretPulledImages` (or something
```

Comment:
This seems hard to maintain overtime, whenever you have a new application using new private image, you need to restart all cluster nodes.

Shouldn't we make this a pod level config?

Reply:
Yes I was thinking pod level..  by processed by I meant in the same way as kublet processes the pull policy in the pod spec.

Reply:
resolved to setting this option at the kubelet scope via flag..